### PR TITLE
Reorganise Account Notification Settings

### PR DESCRIPTION
**Pull request by langleyd**
Implements feature https://github.com/vector-im/element-android/issues/3646

### screenshots
Account Settings > Notifications
image
Default Notifications
image
Mentions and Keywords
image
Other
image

### notes
- Kept the existing settings architecture (`VectorSettingsBaseFragment`, `VectorPreference` `VectorPreferenceCategory`, etc). Changing the architecture to `ViewModels` etc I think would be out of the scope of this PR.
- Introduced `VectorSwitchPreference` to replace `PushRulePreference` and moved all business logic to the fragment level.
- Removing an option (Silent) allowed us to simplify the business logic somewhat but kept implementation functionally equivalent to previous.
- Found some bugs/discrepancies with web(not regressions, they occur in develop), that I will fix as a part of https://github.com/vector-im/element-android/issues/3672.
- Introduced build time resVal to hide V2 until we are ready to introduce this new UI.
- Ran contributing steps/checks
- Update `Global Notifications` to `Default Notifications` on design's advice

### CODE REVIEW

**Review by gaelledel — APPROVED** [review]

**Review by bmarty — COMMENTED** [review]
Wow, IDK why `vector/src/main/java/im/vector/app/core/preference/PushRulePreference.kt` was executable :/

File: vector/src/main/java/im/vector/app/core/preference/VectorCheckboxPreference.kt [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+
+package im.vector.app.core.preference
+
+import android.content.Context
+import android.util.AttributeSet
+import android.widget.TextView
+import androidx.preference.CheckBoxPreference
+import androidx.preference.PreferenceViewHolder
+
+open class VectorCheckboxPreference : CheckBoxPreference {
```
Comment (bmarty) [inline vector/src/main/java/im/vector/app/core/preference/VectorCheckboxPreference.kt:25]: This class does not have to be `open`. It does not have to be sealed as well, but let's keep the default Kotlin behavior (so seal) unless we have to extend the class.

File: vector/src/main/java/im/vector/app/features/settings/VectorSettingsGlobalNotificationPreferenceFragment.kt [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+
+class VectorSettingsGlobalNotificationPreferenceFragment
+    : VectorSettingsPushRuleNotificationPreferenceFragment() {
+
+    override var titleRes: Int = R.string.settings_notification_global
+
+    override val preferenceXmlRes = R.xml.vector_settings_notification_global
+
+    override val prefKeyToPushRuleId: Map<String, String>
+        get() = mapOf(
```
Comment (bmarty) [inline vector/src/main/java/im/vector/app/features/settings/VectorSettingsGlobalNotificationPreferenceFragment.kt:30]: Could be simplified with `override val prefKeyToPushRuleId = mapOf(`

File: vector/src/main/java/im/vector/app/features/settings/VectorSettingsOtherNotificationPreferenceFragment.kt [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+
+    override var titleRes: Int = R.string.settings_notification_other
+
+    override val preferenceXmlRes = R.xml.vector_settings_notification_other
+
+    override val prefKeyToPushRuleId: Map<String, String>
+        get() = mapOf(
+                "SETTINGS_PUSH_RULE_INVITED_TO_ROOM_PREFERENCE_KEY" to RuleIds.RULE_ID_INVITE_ME,
+                "SETTINGS_PUSH_RULE_CALL_INVITATIONS_PREFERENCE_KEY" to RuleIds.RULE_ID_CALL,
+                "SETTINGS_PUSH_RULE_MESSAGES_SENT_BY_BOT_PREFERENCE_KEY" to RuleIds.RULE_ID_SUPPRESS_BOTS_NOTIFICATIONS,
```
Comment (bmarty) [inline vector/src/main/java/im/vector/app/features/settings/VectorSettingsOtherNotificationPreferenceFragment.kt:33]: I'm scared about the `suppress` wording here. So it mean that the check box has the opposite behavior? I know it's in the spec... EDIT: the UI seems to uncheck the box by default, so I guess it's working correctly.

File: vector/src/main/java/im/vector/app/features/settings/VectorSettingsPushRuleNotificationPreferenceFragment.kt [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+import org.matrix.android.sdk.api.pushrules.rest.PushRule
+import org.matrix.android.sdk.api.pushrules.rest.PushRuleAndKind
+
+abstract class VectorSettingsPushRuleNotificationPreferenceFragment
+    : VectorSettingsBaseFragment() {
+
+    /**
+     * @return the bing rule status boolean
+     */
+    fun ruleStatusIndexFor(ruleAndKind: PushRuleAndKind): Boolean {
```
Comment (bmarty) [inline vector/src/main/java/im/vector/app/features/settings/VectorSettingsPushRuleNotificationPreferenceFragment.kt:36]: Please make all the `fun` private where it's possible, default visibility is public in Kotlin

File: vector/src/main/res/values/strings.xml [hunk 5]
```diff
@@ -2946,10 +2965,6 @@
     <string name="error_failed_to_import_keys">Failed to import keys</string>
 
     <string name="settings_notification_configuration">Notifications configuration</string>
-    <string name="settings_messages_at_room">Messages containing @room</string>
-    <string name="settings_messages_in_e2e_one_to_one">Encrypted messages in one-to-one chats</string>
-    <string name="settings_messages_in_e2e_group_chat">Encrypted messages in group chats</string>
-    <string name="settings_when_rooms_are_upgraded">When rooms are upgraded</string>
```
Comment (bmarty) [inline vector/src/main/res/values/strings.xml:2952]: Lint will complain that we have translation but not original string. So either remove all the existing translations or keep the original string and we do the cleanup in the strings from time to time.
I would go for the first option.

File: vector/src/main/res/xml/vector_settings_notification_global.xml [hunk 6]
```diff
@@ -0,0 +1,28 @@
+<?xml version="1.0" encoding="utf-8"?>
+<androidx.preference.PreferenceScreen xmlns:android="http://schemas.android.com/apk/res/android">
+    <im.vector.app.core.preference.VectorPreferenceCategory
+        android:key="SETTINGS_OTHER"
```
Comment (bmarty) [inline vector/src/main/res/xml/vector_settings_notification_global.xml:4]: This is never used, but this is not the correct key (copy/paste error).

**Review by bmarty — APPROVED** [review]
LGMT, some small remarks, but nothing terrific. Thanks!

File: vector/src/main/res/values/strings.xml (hunk 5 above)
Comment (bmarty) [reply vector/src/main/res/values/strings.xml:2952]: My bad, actually those strings has just been moved above. Please ignore my comment

File: vector/src/main/java/im/vector/app/features/settings/notifications/NotificationIndex.kt [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+ * limitations under the License.
+ */
+
+package im.vector.app.features.settings.notifications
+
+import org.matrix.android.sdk.api.pushrules.rest.PushRule
+import org.matrix.android.sdk.api.pushrules.toJson
+
+enum class NotificationIndex(val index: Int) {
+    OFF(0),
```
Comment (ganfra) [inline vector/src/main/java/im/vector/app/features/settings/notifications/NotificationIndex.kt:23]: Enum already have indexes by default starting from 0

**Review by ganfra — APPROVED** [review]
Just one small remark otherwise LGTM

File: vector/src/main/java/im/vector/app/features/settings/notifications/NotificationIndex.kt (hunk 7 above)
Comment (langleyd) [reply vector/src/main/java/im/vector/app/features/settings/notifications/NotificationIndex.kt]: Plus I didn't actually end up using the index value 🥴. Removed, thanks.

**Pull request merged**